bubuku/features/rebalance_by_size.py

```diff
@@ -0,0 +1,120 @@
+import json
+import logging
+
+from kazoo.exceptions import NodeExistsError
+
+from bubuku.broker import BrokerManager
+from bubuku.controller import Check, Change
+from bubuku.utils import CmdHelper
+from bubuku.zookeeper import Exhibitor
+
+_LOG = logging.getLogger('bubuku.features.rebalance_by_size')
+
+
+class RebalanceBySizeChange(Change):
+    def __init__(self, zk: Exhibitor):
+        self.zk = zk
+
+    def get_name(self):
+        return 'rebalance_by_size'
+
+    def can_run(self, current_actions):
+        raise NotImplementedError('Not implemented')  # todo: not implemented yet
+
+    def run(self, current_actions):
+        raise NotImplementedError('Not implemented')  # todo: not implemented yet
+
+    def can_run_at_exit(self):
+        raise NotImplementedError('Not implemented')  # todo: not implemented yet
+
+
+class RebalanceBySize(Check):
+    check_interval_s = 600  # 10 min
+
+    def __init__(self, zk: Exhibitor, broker: BrokerManager):
+        super().__init__()
+        self.zk = zk
+        self.broker = broker
+
+    def check(self):
+        if self.broker.is_running_and_registered() and self.__is_data_imbalanced():
+            _LOG.info("Starting rebalance by size")
+            return RebalanceBySizeChange(self.zk)
+        return None
+
+    def __is_data_imbalanced(self) -> bool:
+        return False  # todo: not implemented yet
+
+
+class GenerateDataSizeStatistics():
+    check_interval_s = 1800  # 30 min
+
+    def __init__(self, zk: Exhibitor, broker: BrokerManager, cmd_helper: CmdHelper, kafka_log_dirs):
+        super().__init__()
+        self.zk = zk
+        self.broker = broker
+        self.cmd_helper = cmd_helper
+        self.kafka_log_dirs = kafka_log_dirs
+
+    def check(self):
+        if self.broker.is_running_and_registered():
+            _LOG.info("Generating data size statistics")
+            try:
+                self.__generate_stats()
+                _LOG.info("Data size statistics successfully written to zk")
+            except Exception:
+                _LOG.warn("Error occurred when collecting size statistics", exc_info=True)
+        return None
+
+    def __generate_stats(self):
+        topic_stats = self.__get_topics_stats()
+        disk_stats = self.__get_disk_stats()
+        stats = {"disk": disk_stats, "topics": topic_stats}
+        self.__write_stats_to_zk(stats)
```

Comment:
I was thinking about the following scenario:

Broker 1 has partitions with size [ 200, 200, 10 ]
Broker 2 has partitions with size [ 300 ]

Both have disk capacity of 1000.

The ideal distribution would be [ 200, 200] and [300, 10 ]. But if we decide to move differently we could have [ 200, 10] and [300, 200] which made it worse.

What I mean is that in order to find the optimal packing of partitions, one should have access to all partitions sizes and do a packing algorithm. Only knowing about the imbalance would not be sufficient for writing an optimal algorithm.

Does it make sense? Writing lots of data might be costly and running such algorithm might be costly as well. Maybe we just want an approximated solution...